File: BestTimeToBuyAndSellStockII.java
// https://www.lintcode.com/problem/150/

// Given an array prices, which represents the price of a stock in each day.

// You may complete as many transactions as you like (ie, buy one and sell one share of the stock multiple times). 

// However, you may not engage in multiple transactions at the same time (ie, if you already have the stock, you must sell it before you buy again).

// Design an algorithm to find the maximum profit.

public class Solution {
    /**
     * @param prices: Given an integer array
     * @return: Maximum profit
     */
    public int maxProfit(int[] prices) {

        int minimum = Integer.MAX_VALUE;

        int result = 0;

        for (int i = 0; i < prices.length; i ++) {

            if (prices[i] - minimum > 0) {
                // 只要有收益 就卖出
                result += prices[i] - minimum;
                // 最小值更新为当前值，而不是Integer.MAX_VALUE
                // 因为这一天也可以买入而不是必须等到下一天
                minimum = prices[i];
            } else {
                minimum = Math.min(minimum, prices[i]);
            }
        }

        return result;
    }
}

File: best_time_to_buy_and_sell_stock_II.py
# https://www.lintcode.com/problem/150/

# Given an array prices, which represents the price of a stock in each day.

# You may complete as many transactions as you like (ie, buy one and sell one share of the stock multiple times). 

# However, you may not engage in multiple transactions at the same time (ie, if you already have the stock, you must sell it before you buy again).

# Design an algorithm to find the maximum profit.
    
# Always sell if you find a profit 

class Solution:
    """
    @param prices: Given an integer array
    @return: Maximum profit
    """
    def maxProfit(self, prices):
        
        if not prices:
            return 0
        
        maximum_profit = 0
        
        # i stands for attempt to sell stock at this day, it does not mean we really sell it.
        # as long as at day i the gain from previous day is postive, we add to the profit
        # this doesn't mean we buy at the previous either, it just mean we CAN hold the stock from yesterday until today
        # we only sell when the gain is negative
        for i in range(1, len(prices)):
            if prices[i] - prices[i-1] > 0:
                maximum_profit += (prices[i] - prices[i-1])
        
        return maximum_profit
